src/mlir/design.md

```diff
@@ -3,69 +3,225 @@
+
+Lambdas will be lowered as an anonymous structure with captures as fields and `apply()` as the method for when the lambda is called.
+That transformation should have already been done by the parser and represented in the AST by the right basic constructs.
+At the MLIR level, they'll be allocated and dispatched in the usual way (stack/heap + call/behaviour dispatch).
+
+**QUESTION: How do we mark the `main` function?**
+Currently, if there is a function called `main`, it's lowered without scope to help testing execution, but we need more strict rules, or perhaps annotation, to mark the entry point.
+We could also have a class named `Main` which is restricted to only one per program, with a method `main(int, int**) : int`, which is the entry point.
+We then create a compiler-generated entry point and call that method with the command line arguments, redirecting the return value.
+
+## Runtime Semantics
+
+Cowns, behaviours and regions are created and executed by the runtime.
+At compile time, all we need to do is to make sure we call the right runtime functions.
+
+Memory allocation will be selected by the type of the regions, which will indicate which `snmalloc` functions we'll call for each of their objects.
+Cown reservation and behaviour execution (`when` blocks) will become a call to the appropriate scheduler functions with references to the right objects.
+
+The parser will have already done liveness analysis, so we know when variables (including heap-allocated regions) go out of scope.
+It will also have introduced nodes for their deallocations, which will be converted into calls to `snmalloc` or garbage collector hints.
 
 ## Load/Store Semantics
 
 Stack variables are allocated by LLVM IR with an `alloca` instruction.
 MLIR doesn't support allocating stack memory with anything other than `memref` types, but it does support the LLVM dialect, which has full LLVM semantics.
-In this preliminary implementation, we use the LLVM dialect for alloca/load/store/GEP operations when dealing with stack objects.
-This approach seems very clean and sensible, and it could be used for accessing heap variables as well.
-
-For now, every function argument and local declaration (`let`, `var`) are allocated in the stack and its value stored there.
-This simplifies code generation (all variable symbols point to an address), but it doesn't help with value symbols and are really slow.
+We use the LLVM dialect for alloca/load/store/GEP operations when dealing with stack/heap objects.
 
-The LLVM pipeline, even at O1, recognises most of the argument, temporary and local allocas as bogus and optimises them away, so this isn't an immediately critical issue from a performance point of view, but we will want to clean up some of those repetitive patterns soon enough.
+Local declarations (`let`, `var`) without region information are allocated in the stack, via LLVM's `alloca` instruction and their types.
```

Comment:
This presumably means that we'll need to have a later pass to promote some of these to the heap if they are very large or if they escape.  The converse approach (which I think the Go compiler uses) is to model these in the early IR as a heap allocations and then to have a later pass that converts them to stack allocations where possible.